types/validator_set.go

```diff
@@ -63,27 +64,46 @@ func (vals *ValidatorSet) CopyIncrementAccum(times int) *ValidatorSet {
 // IncrementAccum increments accum of each validator and updates the
 // proposer. Panics if validator set is empty.
 func (vals *ValidatorSet) IncrementAccum(times int) {
+	// call IncrementAccum(1) times times:
+	for i := 0; i < times; i++ {
+		vals.incrementAccum()
+	}
+}
 
-	// Add VotingPower * times to each validator and order into heap.
+func (vals *ValidatorSet) incrementAccum() {
```

Comment:
I wonder if we really need to average every time `incrementAccum()` is called, that seems a bit wasteful - presumably we could just average at the end of `IncrementAccum` as long as we don't call it too many times in a row (or average every `n` calls).

Reply:
Agreed but wouldn't his contradict #2809:

> IncrementAccum() was written to efficiently calculate the accum as if IncrementAccum(1) was called times times

Calling it only every n times would break this invariant, too, or not? Most of the times we call IncrementAccum with times = 1 and we wouldn't average in most calls.

Reply:
If we average at the end of `IncrementAccum()`, as long as there are no overflows, the invariant should be maintained.

Reply:
That's true. Thanks! I'm not sure this optimization is worth the additional code complexity 🤔 
I'll add this logic and leave it for review.

Reply:
https://github.com/tendermint/tendermint/pull/2785/commits/c98841c2574b0bedff0528384606ce95339365c9